Add RMC as a muon stop pileup product option

michaelmackenzie opened the pull request:
Add logic for RMC to be included as a muon stop pileup product. This mostly entailed adding a flag to the RMC tool to use the RMC rate per muon to generate a random number of photons (usually 0), and add the corresponding global constant. The global constant RMC rate for aluminum is 1.4e-5 for E > 57 MeV over the closure approximation spectrum fraction for E > 57 MeV (~1/10,000 muon captures). I will make a corresponding PR in Production adding this product to the MuStopPileup configuration.

FNALbuild commented:
:sunny: The build tests passed at e944805bc60a8dc798f2d73d7be4773ca8f20c94.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged e944805bc60a8dc798f2d73d7be4773ca8f20c94 at f5aa30e4061fa8ca01c61d5c03157ef47771932a |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/scons.log). Build time: 04 min 18 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 3 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 120 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at e944805bc60a8dc798f2d73d7be4773ca8f20c94 after being merged into the base branch at f5aa30e4061fa8ca01c61d5c03157ef47771932a.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2320/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.

brownd1978 reviewed (changes requested):
I just realized that this tool is using unvalidated fhicl. We have the policy that all (legacy) codes using unvalidated fhicl must upgrade when they are modified. Validated fhicl examples can be found throughout our code base, look for ‘Config’ structs.

brownd1978 commented on EventGenerator/src/RMCGenerator_tool.cc:
```diff
@@ -45,14 +45,16 @@ namespace mu2e {
       _czmax(conf().spectrum.get<fhicl::ParameterSet>().get<double>("czmax",  1.)),
       _spectrum(BinnedSpectrum(conf().spectrum.get<fhicl::ParameterSet>())),
+      _useRate(conf().spectrum.get<fhicl::ParameterSet>().get<bool>("useRate",  false)),
+      _internalRate((_external) ? 0. : 1.),
```
I’m struggling to follow the settings needed to get the following configurations
- external (photons) only
- internal conversions only
- physical admixture of internal and external
Some combination of the ‘external’ and ‘userate’ bools seems to provide these, but I can’t quite follow.
It would be clearer if the valid configuration options were explicitly enumerated (enum) and set in the config.

I also find the term ‘external’ for producing RMC photons unintuitive, as those photons might be detected directly in the calorimeter, in which case there’s nothing ‘external’ about them.

michaelmackenzie replied:
Only one of the `external` or `useRate` flags are necessary. I updated the configuration to use validated fhicl and in the update I set the `external` flag to only be required if `useRate` is false. I could update this to use strings or enums to check for something like "natural", "external", or "internal" if this is a preferred logic. I also don't prefer `external` as I term (I typically write real photons or virtual photons/internal conversions in talks), but it better matches the process IDs used and alternate generators.

brownd1978 replied:
I prefer the conceptually more clear config to choose between the 3 sensible options. The names you give above are clear. If the process names are unclear we can change them, but lets do that in a separate PR.

brownd1978 commented on GlobalConstantsService/data/globalConstants_01.txt:
```diff
@@ -125,6 +125,7 @@ double physicsParams.Al.capture.protonRate = 0.05;
 double physicsParams.Al.capture.deuteronRate = 0.025;
 double physicsParams.Al.capture.neutronRate = 1.2;
 double physicsParams.Al.capture.photonRate = 2.0;
+double physicsParams.Al.capture.RMCRate = 9.460e-5;
```
Please add it as a comment  after this line. We need  to document the provenance of every physics number in this table, which is beyond the scope of this PR. I will open an issue once the PR is merged

michaelmackenzie replied:
I added a note here and for the internal RMC branching fraction to reference where it comes from.

michaelmackenzie commented:
I've updated the `RMCGenerator_tool` to use three modes: `external`, `internal`, or `physical`. I've updated the corresponding Production PR. This will require additional updates to the Production RMC primary fcl, but unfortunately I branched off Production before the RMC primary fcl were merged as I didn't anticipate changing how the tool is configured for primary productions. Is it best to pull in the Production updates (or rebase), or should I make a new PR for this fix after these updates are merged?

brownd1978 commented:
Hi Michael,
thanks for changing the config options.
Merging (or rebasing) to the current head in your work branch is the
suggested workflow for this situation. The PR will follow those changes
once you commit and push.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for 4a65982c85b05d78be745821121675b847452349: build  (Build queue is empty)

FNALbuild commented:
:sunny: The build tests passed at 4a65982c85b05d78be745821121675b847452349.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| test with | :white_check_mark: | Command did not list any other PRs to include |
| merge | :white_check_mark: | Merged 4a65982c85b05d78be745821121675b847452349 at 2ff5bb38d9af3c9ac2c1bc4a7acaab92313b61f0 |
| build (prof) | :white_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/scons.log). Build time: 04 min 17 sec |
| ceSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/ceSimReco.log)  |
| g4test_03MT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/g4test_03MT.log)  |
| transportOnly | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/transportOnly.log)  |
| POT | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/POT.log)  |
| g4study | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/g4study.log)  |
| cosmicSimReco | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/cosmicSimReco.log)  |
| cosmicOffSpill | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/cosmicOffSpill.log)  |
| ceSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/ceSteps.log)  |
| ceDigi | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/ceDigi.log)  |
| muDauSteps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/muDauSteps.log)  |
| ceMix | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/ceMix.log)  |
| rootOverlaps | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/rootOverlaps.log)  |
| g4surfaceCheck | :white_check_mark: | [Log file.](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/g4surfaceCheck.log)  |
| FIXME, TODO | :white_check_mark: | [TODO (0) FIXME (0) in 3 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/fixme_todo.log) |
| clang-tidy | :large_orange_diamond: | [0 errors 122 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/clang-tidy.log) |
| whitespace check | :white_check_mark: | [no whitespace errors found](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/artifact/whitespace_errs.log) |


N.B. These results were obtained from a build of this Pull Request at 4a65982c85b05d78be745821121675b847452349 after being merged into the base branch at 2ff5bb38d9af3c9ac2c1bc4a7acaab92313b61f0.


For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//2332/console).
Build artifacts are deleted after 5 days. If this is not desired, select `Keep this build forever` on the job page.
